File: Week2/D4.java
import java.util.*;

class Week2_4 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt(), k = sc.nextInt();

        String[][] land = new String[n][n];
        int[][] score = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                land[i][j] = sc.next();
                score[i][j] = 0;
            }
        }

        for (int i = 0; i < k; i++) {
            int y = sc.nextInt() - 1, x = sc.nextInt() - 1;
            for (int a = x - 1; a < x + 2; a++) {
                try {
                    String tile = land[y][a];
                    if (tile.equals("@"))
                        score[y][a] += 2;
                    else if (tile.equals("0"))
                        score[y][a]++;
                } catch (ArrayIndexOutOfBoundsException e) {
                    continue;
                }
            }
            for (int b = y - 1; b < y + 2; b += 2) {
                try {
                    String tile = land[b][x];
                    if (tile.equals("@"))
                        score[b][x] += 2;
                    else if (tile.equals("0"))
                        score[b][x]++;
                } catch (ArrayIndexOutOfBoundsException e) {
                    continue;
                }
            }
        }

        int result = 0;
        for (int[] a : score) {
            for (int b : a) {
                if (b > result)
                    result = b;
            }
        }

        System.out.println(result);
    }
}



/**
코드 해설

1. 한 변 길이(이하 n), 폭탄 떨어트릴 횟수(이하 k) 입력
2. n * n 크기의 2차원 String 배열(해당 칸 상태 저장용, 이하 land)와 int 배열(해당 칸 점수 저장용, 이하 score) 생성
3. 폭탄 떨어트리기
	1) 입력한 좌표에 해당하는 land 값과 상하좌우의 값을 확인
	2) 숫자면 입력한 좌표에 해당하는 score 값 + 1, @면 +2, #이면 아무것도 하지 않음
	3) 만약 인덱스 오류가 발생하면 try-catch로 패스
4. score 배열 순회, 최댓값 출력
 */

File: Week2/D4.py
def bomb(tile) :
    try :
        return str(int(tile) + 1)
    except ValueError :
        if tile[0] == '@' : return tile + '!'
        else : return tile

n, k = map(int, input().split())

land = [['0' for _ in range(n + 2)]] + [['0'] + list(map(str, input().split())) + [0] for _ in range(n)] + [['0' for _ in range(n + 2)]]

for _ in range(k) :
    y, x = map(int, input().split())
    for i in range(x - 1, x + 2) :
        land[y][i] = bomb(land[y][i])
    for j in range(y - 1, y + 2, 2) :
        land[j][x] = bomb(land[j][x])

result = 0
for li in land[1:n + 1] :
    for sq in li[1:n + 1] :
        if sq[0] == "@" :
            temp = sq.count('!') * 2
            if temp > result : result = temp
        elif sq == "#" : continue
        else :
            if int(sq) > result : result = int(sq)
            
print(result)


"""
코드 해설

1. 한 변 길이(이하 n), 떨어트릴 폭탄 개수(이하 k) 입력
2. (n + 2) * (n + 2) 크기의 2차원 배열(이하 land) 생성
3. 가장자리를 제외하고 n * n 구역의 픽셀 값 입력
4. 폭탄을 떨어트리는 함수 bomb 실행
    1) 해당 좌표와 그 좌표의 상하좌우 위치의 값을 바꿈
    2) 숫자면 해당 값 + 1, @면 문자열 뒤에 ! 추가, #면 아무것도 하지 않고 값 반환
5. 배열 탐색하여 최대값 계산(숫자면 그대로, @면 (!의 개수) * 2, #면 계산하지 않음) 후 출력
"""